File: modules/swagger-core/src/test/scala/1_3/converter/SnakeCaseConverterTest.scala
package converter

import models._

import com.wordnik.swagger.converter._
import com.wordnik.swagger.models._
import com.wordnik.swagger.models.properties._

import com.wordnik.swagger.annotations._
import com.wordnik.swagger.converter._
import com.wordnik.swagger.util.Json
import com.wordnik.swagger.models._

import java.lang.annotation.Annotation
import java.lang.reflect.Type
import java.util.LinkedHashMap
import javax.xml.bind.annotation._

import scala.beans.BeanProperty
import scala.collection.JavaConverters._

import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.FlatSpec
import org.scalatest.Matchers

import matchers.SerializationMatchers._

@RunWith(classOf[JUnitRunner])
class SnakeCaseConverterTest extends FlatSpec with Matchers {
  it should "ignore properties with type Bar" in {
    // add the custom converter
    val snakeCaseConverter = new SnakeCaseConverter
    val converters = new ModelConverters()

    converters.addConverter(snakeCaseConverter)

    val models = converters.readAll(classOf[SnakeCaseModel])
    models should serializeToJson (
"""{
  "bar" : {
    "properties" : {
      "foo" : {
        "type" : "string"
      }
    }
  },
  "snake_case_model" : {
    "properties" : {
      "bar" : {
        "$ref" : "#/definitions/bar"
      },
      "title" : {
        "type" : "string"
      }
    },
    "xml" : {
      "name" : "snakeCaseModel"
    }
  }
}""")
  }
}

@XmlRootElement(name="snakeCaseModel")
class SnakeCaseModel {
  @BeanProperty var bar:Bar = null
  @BeanProperty var title:String = null
}

/**
 * simple converter to rename models and field names into snake_case
 */
class SnakeCaseConverter extends ModelConverter {
  def resolveProperty(`type`: Type, context: ModelConverterContext, annotations: Array[Annotation], chain: java.util.Iterator[ModelConverter]): Property = {
    if(chain.hasNext()) {
      val converter = chain.next()
      return converter.resolveProperty(`type`, context, annotations, chain)
    }
    return null
  }

  def resolve(`type`: Type, context: ModelConverterContext, chain: java.util.Iterator[ModelConverter]): Model = {
    if(chain.hasNext()) {
      val converter = chain.next()
      val model = converter.resolve(`type`, context, chain)
      if(model != null) {
        val properties = model.getProperties()
        val updatedProperties = new LinkedHashMap[String, Property]
        for(key <- properties.keySet.asScala) {
          val convertedKey = toSnakeCase(key)
          val prop = properties.get(key)
          if(prop.isInstanceOf[RefProperty]) {
            val ref = prop.asInstanceOf[RefProperty]
            ref.set$ref(toSnakeCase(ref.getSimpleRef()));
          }
          updatedProperties.put(convertedKey, prop)
        }
        model.getProperties().clear()
        model.setProperties(updatedProperties)
        val name = (model.asInstanceOf[ModelImpl]).getName
        if(model.isInstanceOf[ModelImpl]) {
          val impl = model.asInstanceOf[ModelImpl]
          impl.setName(toSnakeCase(impl.getName()))
        }
        return model
      }
    }
    return null
  }

  val primitives = Set("string", "integer", "number", "boolean", "long")
  def toSnakeCase(str: String) = {
    val o = "[A-Z\\d]".r.replaceAllIn(str, {m => "_" + m.group(0).toLowerCase()})
    if(o.startsWith("_")) o.substring(1)
    else o
  }
}

File: modules/swagger-core/src/test/java/io/swagger/model/override/SnakeCaseConverterTest.java
package io.swagger.model.override;

import com.google.common.collect.Sets;
import io.swagger.converter.ModelConverter;
import io.swagger.converter.ModelConverterContext;
import io.swagger.converter.ModelConverters;
import io.swagger.matchers.SerializationMatchers;
import io.swagger.oas.models.media.Schema;
import io.swagger.util.Json;
import org.testng.annotations.Test;

import javax.xml.bind.annotation.XmlRootElement;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

public class SnakeCaseConverterTest {

    @Test(description = "it should change naming style")
    public void testConvert() {
        // add the custom converter
        final SnakeCaseConverter snakeCaseConverter = new SnakeCaseConverter();
        final ModelConverters converters = new ModelConverters();

        converters.addConverter(snakeCaseConverter);

        final Map<String, Schema> models = converters.readAll(SnakeCaseModel.class);
        final String json = "{" +
                "   \"bar\":{" +
                "      \"title\" : \"bar\",\n" +
                "      \"type\":\"object\"," +
                "      \"properties\":{" +
                "         \"foo\":{" +
                "            \"title\" : \"foo\"," +
                "            \"type\":\"string\"" +
                "         }" +
                "      }" +
                "   }," +
                "   \"snake_case_model\":{" +
                "      \"type\":\"object\"," +
                "      \"properties\":{" +
                "         \"bar\":{" +
                "            \"$ref\":\"#/definitions/bar\"" +
                "         }," +
                "         \"title\":{" +
                "            \"type\":\"string\"" +
                "         }" +
                "      }," +
                "      \"xml\":{" +
                "         \"name\":\"snakeCaseModel\"" +
                "      }" +
                "   }" +
                "}";
        Json.prettyPrint(models);
        SerializationMatchers.assertEqualsToJson(models.get("snake_case_model"), json);
    }

    @XmlRootElement(name = "snakeCaseModel")
    class SnakeCaseModel {
        public Bar bar = null;
        public String title = null;
    }

    /**
     * simple converter to rename models and field names into snake_case
     */
    class SnakeCaseConverter implements ModelConverter {
        final Set<String> primitives = Sets.newHashSet("string", "integer", "number", "boolean", "long");

        @Override
        public Schema resolve(Type type, ModelConverterContext context, Annotation[] annotations, Iterator<ModelConverter> chain) {
            if (chain.hasNext()) {
                final ModelConverter converter = chain.next();
                return converter.resolve(type, context, annotations, chain);
            }
            return null;
        }

        @Override
        public Schema resolve(Type type, ModelConverterContext context, Iterator<ModelConverter> chain) {
            if (chain.hasNext()) {
                final ModelConverter converter = chain.next();
                final Schema model = converter.resolve(type, context, chain);
                if (model != null) {
                    final Map<String, Schema> properties = model.getProperties();
                    final Map<String, Schema> updatedProperties = new LinkedHashMap<String, Schema>();
                    for (String key : properties.keySet()) {
                        String convertedKey = toSnakeCase(key);
                        Schema prop = properties.get(key);
                        if (prop.get$ref() != null) {
                            // TODO
//                            prop.set$ref(toSnakeCase(ref.getSimpleRef()));
                        }
                        updatedProperties.put(convertedKey, prop);
                    }
                    model.getProperties().clear();
                    model.setProperties(updatedProperties);
                    if (model instanceof Schema) {
                        Schema impl = (Schema) model;
                        String prevName = impl.getTitle();
                        impl.setTitle(toSnakeCase(impl.getTitle()));
                        context.defineModel(impl.getTitle(), impl, type, prevName);
                    }
                    return model;
                }
            }
            return null;
        }

        private String toSnakeCase(String str) {
            String o = str.replaceAll("[A-Z\\d]", "_"+ "$0").toLowerCase();
            if (o.startsWith("_")) {
                return o.substring(1);
            } else {
                return o;
            }
        }
    }

    class Bar {
        public String foo = null;
    }
}
